processing/src/main/java/org/apache/druid/guice/ExtensionsLoader.java

```diff
@@ -187,25 +201,84 @@ public File[] getExtensionFilesToLoad()
         extensionsToLoad[i++] = extensionDir;
       }
     }
-    return extensionsToLoad == null ? new File[]{} : extensionsToLoad;
+    extensionFilesToLoad = extensionsToLoad == null ? new File[]{} : extensionsToLoad;
+  }
+
+  public File[] getExtensionFilesToLoad()
+  {
+    if (extensionFilesToLoad == null) {
+      initializeExtensionFilesToLoad();
+    }
+    return extensionFilesToLoad;
   }
 
   /**
    * @param extension The File instance of the extension we want to load
    *
    * @return a URLClassLoader that loads all the jars on which the extension is dependent
    */
-  public URLClassLoader getClassLoaderForExtension(File extension, boolean useExtensionClassloaderFirst)
+  public StandardURLClassLoader getClassLoaderForExtension(File extension, boolean useExtensionClassloaderFirst)
   {
-    return loaders.computeIfAbsent(
-        Pair.of(extension, useExtensionClassloaderFirst),
-        k -> makeClassLoaderForExtension(k.lhs, k.rhs)
-    );
+    return getClassLoaderForExtension(extension, useExtensionClassloaderFirst, new ArrayList<>());
+  }
+
+  public StandardURLClassLoader getClassLoaderForExtension(File extension, boolean useExtensionClassloaderFirst, List<String> extensionDependencyStack)
```

Comment:
Javadoc please, explaining what `extensionDependencyStack` is and how it's used. It isn't obvious from the context.